worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionFilesSorter.java

```diff
@@ -687,4 +709,88 @@ private void readBufferBySize(ByteBuffer buffer, int toRead) throws IOException
       }
     }
   }
+
+  public void cleanupExpiredShuffleKey(Set<String> expiredShuffleKeys) {
+    cleaner.cleanupExpiredShuffleKey(expiredShuffleKeys);
+  }
+}
+
+class PartitionFilesCleaner {
+  private static final Logger logger = LoggerFactory.getLogger(PartitionFilesCleaner.class);
+
+  private final LinkedBlockingQueue<PartitionFilesSorter.FileSorter> queue =
+      new LinkedBlockingQueue<>();
+  private final Lock lock = new ReentrantLock();
+  private final Condition notEmpty = lock.newCondition();
+
+  PartitionFilesCleaner(ChunkStreamManager chunkStreamManager) {
+    Thread cleaner =
+        new Thread(
+            () -> {
+              while (true) {
+                try {
+                  lock.lockInterruptibly();
+                  while (queue.isEmpty()) {
+                    notEmpty.await(500, TimeUnit.MILLISECONDS);
+                  }
+                  Iterator<PartitionFilesSorter.FileSorter> it = queue.iterator();
+                  while (it.hasNext()) {
+                    PartitionFilesSorter.FileSorter sorter = it.next();
+                    boolean streamFetchFinished = true;
+                    Set<Long> streamIds =
+                        chunkStreamManager.shuffleStreamIds.get(sorter.getShuffleKey());
+
+                    if (streamIds != null) {
+                      for (long streamId : streamIds) {
+                        ChunkStreamManager.StreamRegisterState registerState =
+                            chunkStreamManager.streamRegisterStates.get(streamId);
+                        if (registerState != null) {
+                          boolean hasOriginalFileFetcher =
+                              !registerState.isRangeRead()
+                                  // TODO:(fchen)
+                                  // && registerState.isReadBufferFromOriginalFile()
+                                  && registerState.isRegisterBefore(sorter.getStartTimestamp());
```

Comment:
Seems original fetch can be registered after sorter's timestamp. For example at time1 a range reader comes and `FileSorter` is created, at time2 a non-range reader comes and at this time the sort task has not finished, so it uses original file but has larger timestamp than `FileSorter`'s.